chemprop/cli/hyperopt.py

```diff
@@ -1,16 +1,348 @@
+import json
+import logging
+import sys
 from argparse import ArgumentParser, Namespace
+from pathlib import Path
 
+import torch
+from lightning import pytorch as pl
+from lightning.pytorch.loggers import CSVLogger, TensorBoardLogger
+from lightning.pytorch.callbacks import ModelCheckpoint, EarlyStopping
+from ray import tune
+from ray.train import CheckpointConfig, RunConfig, ScalingConfig
+from ray.train.lightning import (RayDDPStrategy, RayLightningEnvironment,
+                                 RayTrainReportCallback, prepare_trainer)
+from ray.train.torch import TorchTrainer
+from ray.tune.schedulers import ASHAScheduler
+
+from chemprop.cli.common import (add_common_args, process_common_args,
+                                 validate_common_args)
+from chemprop.cli.train import (add_train_args, build_datasets, build_model,
+                                build_splits, normalize_inputs,
+                                process_train_args, validate_train_args)
 from chemprop.cli.utils.command import Subcommand
+from chemprop.data import MolGraphDataLoader
+from chemprop.nn import AggregationRegistry
+from chemprop.nn.utils import Activation
+
+
+logger = logging.getLogger(__name__)
+
+AVAILABLE_SPACES = {
+    "activation": tune.choice(list(Activation.keys())),
+    "aggregation": tune.choice(list(AggregationRegistry.keys())),
+    "aggregation_norm": tune.quniform(lower=1, upper=200, q=1),
+    "batch_size": tune.quniform(lower=5, upper=200, q=5),
+    "depth": tune.randint(lower=2, upper=6),
+    "dropout": tune.choice(
+        [
+            tune.choice([0.]),
+            tune.quniform(lower=0.05, upper=0.4, q=0.05),
+        ],
+    ),
+    "ffn_hidden_size": tune.quniform(lower=300, upper=2400, q=100),
+    "ffn_num_layers": tune.randint(lower=2, upper=6),
+    "final_lr_ratio": tune.loguniform(lower=1e-4, upper=1),
+    "hidden_size": tune.quniform(lower=300, upper=2400, q=100),
+    "init_lr_ratio": tune.loguniform(lower=1e-4, upper=1),
+    "linked_hidden_size": tune.quniform(lower=300, upper=2400, q=100),
```

Comment:
I'm a little confused by `linked_hidden_size`. From what I can see it changes both `hidden_size` and `ffn_hidden_size` together, but these are both already in the search space themselves. Would sampling both of these parameters not already cover what `linked_hidden_size` does?

Reply:
If the user only specify `linked_hidden_size` then it will optimize both, but if the user specify `hidden_size` and `ffn_hidden_size` separately, it will optimize them separately. I try to recover the v1 behavior here

Reply:
Not all available spaces are optimized. Only the ones specificed by the users are.

Reply:
The purpose of using `linked_hidden_size` would be to reduce the dimension of the search space by 1, as mentioned in the original #299 PR.

Reply:
@am2145 This limits the search space and the user can still explicitly ask these two to be searched separately.